## IVTR Lead: FAIL — REQUEST CHANGES — 2 blocking failures

Independent validation of commit `c8419c403` against T9814 ACs. Tests pass (6/6) but 2 blocking issues found.

### FAIL #1 — AC2: Transaction semantic violation (BLOCKING)

**AC2 states**: "Single dataAccessor.transaction wraps all inserts — verified by reading the implementation"

**Observed**: The implementation does NOT use a single wrapping `dataAccessor.transaction()`. Instead it uses a compensating-delete (saga) pattern:
- Each `addTask` call runs in its own internal transaction  
- On failure, a separate `compensatingDelete()` deletes already-created tasks in a second transaction
- There is a small window between the last successful insert and the cleanup where rows are visible in the DB

**Location**: `packages/core/src/tasks/add-batch.ts:128-168` (live path) and `compensatingDelete` at line 188

**Note**: The file's own JSDoc acknowledges this is architecturally constrained (allocateNextTaskId uses raw BEGIN IMMEDIATE that cannot be nested inside a DataAccessor transaction). However, AC2 explicitly requires the wrapping-transaction approach. Either the AC must be updated to match the compensating-delete design, OR the implementation must be rearchitected.

**Required action**: Either (a) implement true single-transaction wrapping (may require refactoring allocateNextTaskId), OR (b) coordinate with the orchestrator to update AC2 to reflect the compensating-delete approach.

### FAIL #2 — TypeScript type error in dispatch wiring (BLOCKING — CI confirmed FAILURE)

**CI check**: "Type Check" FAILED on GitHub Actions run 26207641329

**Error** (`packages/cleo/src/dispatch/domains/tasks.ts:298`):
```
error TS2345: Argument of type 'AddBatchResult' is not assignable to parameter of type '{ success: boolean; ... }'.
Property 'success' is missing in type 'AddBatchResult' but required.
```

**Root cause**: `tasksAddBatchOp` returns raw `AddBatchResult = { created, tasks, dryRun }` — no `success` field. But `wrapCoreResult` expects `EngineResult`-shaped `{ success: boolean; ... }`. All other ops (add, update, etc.) use `engineSuccess`/`engineError` wrappers.

**Fix**: Wrap the result in `engineSuccess(result)` before passing to `wrapCoreResult`, OR refactor `tasksAddBatchOp` to return `EngineResult<AddBatchResult>`.

### PASS items

- AC1: `packages/core/src/tasks/add-batch.ts` exists, exports `addBatchTasks` ✓
- AC3: capability-matrix entry `{ domain: 'tasks', operation: 'add-batch', gateway: 'mutate' }` at lines 152-158 ✓
- AC4: Dispatch handler registered in `packages/cleo/src/dispatch/domains/tasks.ts:295` ✓ (blocked by type error)
- AC5: Unit test (a) happy path 3 tasks — PASS ✓
- AC6: Unit test (b) 2nd task bad parent → ZERO rows confirmed ✓
- AC7: Unit test (c) 3rd task rollback → ZERO rows confirmed ✓
- AC8: `pnpm --filter @cleocode/core run test --run add-batch` → 6/6 pass ✓
- AC9: Code in `packages/core/`, not `packages/cleo/` ✓

### Verdict

2 blocking ACs fail. Task remains pending. Fix FAIL #2 (type error) first, then resolve FAIL #1 (AC2 transaction semantic).

--- a/packages/cleo/src/dispatch/domains/tasks.ts
+++ b/packages/cleo/src/dispatch/domains/tasks.ts
@@ -71,6 +71,7 @@ import {
   taskSyncLinks,
   taskSyncLinksRemove,
   taskSyncReconcile,
+  tasksAddBatchOp,
   taskTree,
   taskUnarchive,
   taskUnclaim,
@@ -290,6 +291,18 @@ const _tasksTypedHandler = defineTypedHandler<TasksOps>('tasks', {
   // -------------------------------------------------------------------------
   // Mutate ops
   // -------------------------------------------------------------------------
+
+  'add-batch': async (params) => {
+    const projectRoot = getProjectRoot();
+    return wrapCoreResult(
+      await tasksAddBatchOp(projectRoot, {
+        tasks: (params.tasks ?? []) as Parameters<typeof tasksAddBatchOp>[1]['tasks'],
+        defaultParent: typeof params.defaultParent === 'string' ? params.defaultParent : undefined,
+        dryRun: typeof params.dryRun === 'boolean' ? params.dryRun : undefined,
+      }),
+      'add-batch',
+    );
+  },
 
   add: async (params) => {
     const projectRoot = getProjectRoot();
@@ -560,6 +573,7 @@ const QUERY_OPS = new Set<string>([
 
 const MUTATE_OPS = new Set<string>([
   'add',
+  'add-batch',
   'update',
   'complete',
   'cancel',

--- a/packages/cleo/src/dispatch/lib/engine.ts
+++ b/packages/cleo/src/dispatch/lib/engine.ts
@@ -261,6 +261,7 @@ export {
   taskSyncLinks,
   taskSyncLinksRemove,
   taskSyncReconcile,
+  tasksAddBatchOp,
   taskTree,
   taskUnarchive,
   taskUnclaim,

--- a/packages/contracts/src/index.ts
+++ b/packages/contracts/src/index.ts
@@ -1037,6 +1037,8 @@ export type {
   DepGraphIssue,
   DepsTreeEdge,
   DepsTreeNode,
+  TasksAddBatchParams,
+  TasksAddBatchResult,
   TasksAddParams,
   TasksAddResult,
   TasksAnalyzeQueryParams,

--- a/packages/contracts/src/operations/tasks.ts
+++ b/packages/contracts/src/operations/tasks.ts
@@ -653,6 +653,63 @@ export interface TasksRelatesRemoveResult {
   removed: boolean;
 }
 
+// tasks.add-batch (atomic multi-task insert)
+/**
+ * Parameters for `tasks.add-batch`.
+ *
+ * Wraps N `tasks.add` inserts in a single transaction; any failure rolls
+ * back ALL inserts. Closes the partial-batch bug in the CLI `add-batch`
+ * command (T9813 / T9814).
+ *
+ * @task T9814
+ */
+export interface TasksAddBatchParams {
+  /**
+   * Array of task specs to insert. Must be non-empty.
+   * Each spec accepts the same fields as `tasks.add` (except `dryRun`
+   * which is hoisted to the batch level).
+   */
+  tasks: Array<{
+    title: string;
+    description?: string;
+    parent?: string;
+    depends?: string[];
+    priority?: string;
+    labels?: string[];
+    type?: string; // SSoT-EXEMPT:kind≠type — 'type' is hierarchy(epic|task|subtask), 'kind' is intent(work|bug|...) — separate axes T944
+    acceptance?: string[];
+    phase?: string;
+    size?: string;
+    notes?: string;
+    files?: string[];
+    kind?: string;
+    scope?: string;
+    severity?: string;
+    forceDuplicate?: boolean;
+  }>;
+  /** Optional default parent ID applied when a task spec omits `parent`. */
+  defaultParent?: string;
+  /**
+   * Dry-run mode: validate each spec and return predicted IDs without
+   * writing to the database. Created count is always 0 in dry-run.
+   */
+  dryRun?: boolean;
+}
+
+/**
+ * Result of `tasks.add-batch`.
+ *
+ * @task T9814
+ */
+export interface TasksAddBatchResult {
+  /** Number of tasks actually created (0 on rollback or dry-run). */
+  created: number;
+  /** Per-task results in input order. */
+  tasks: TasksAddResult[];
+  /** Whether this was a dry run. */
+  dryRun?: boolean;
+}
+
 // tasks.add (dispatch-level params — extends TasksCreateParams)
 export interface TasksAddParams {
   title: string;
@@ -1043,6 +1100,7 @@ export type TasksOps = {
   readonly 'sync.links': readonly [TasksSyncLinksParams, TasksSyncLinksResult];
   // Mutate ops
   readonly add: readonly [TasksAddParams, TasksAddResult];
+  readonly 'add-batch': readonly [TasksAddBatchParams, TasksAddBatchResult];
   readonly update: readonly [TasksUpdateQueryParams, TasksUpdateQueryResult];
   readonly complete: readonly [TasksCompleteQueryParams, TasksCompleteQueryResult];
   readonly cancel: readonly [TasksCancelParams, TasksCancelResult];

--- a/packages/core/src/internal.ts
+++ b/packages/core/src/internal.ts
@@ -1748,6 +1748,8 @@ export type {
   Resolution,
 } from './store/restore-json-merge.js';
 export { regenerateAndCompare, regenerateAndCompareAll } from './store/restore-json-merge.js';
+// Batch task creation with single-transaction atomicity (T9814)
+export { tasksAddBatchOp } from './tasks/add-batch.js';
 export { taskArchive } from './tasks/archive.js';
 export {
   completeTaskStrict,

--- a/packages/core/src/routing/capability-matrix.ts
+++ b/packages/core/src/routing/capability-matrix.ts
@@ -151,6 +151,13 @@ const CAPABILITY_MATRIX: OperationCapability[] = [
   { domain: 'tasks', operation: 'add', gateway: 'mutate', mode: 'native', preferredChannel: 'cli' },
   {
     domain: 'tasks',
+    operation: 'add-batch',
+    gateway: 'mutate',
+    mode: 'native',
+    preferredChannel: 'cli',
+  },
+  {
+    domain: 'tasks',
     operation: 'update',
     gateway: 'mutate',
     mode: 'native',

--- a/packages/core/src/sequence/index.ts
+++ b/packages/core/src/sequence/index.ts
@@ -292,14 +292,22 @@ const MAX_ALLOC_RETRIES = 3;
 /**
  * Atomically allocate the next task ID via SQLite.
  *
- * Uses BEGIN IMMEDIATE to guarantee no two concurrent callers
- * receive the same ID, even across processes (WAL mode).
+ * Uses a SAVEPOINT (instead of BEGIN IMMEDIATE) so that this function
+ * can be called from within an outer transaction without causing
+ * "cannot start a transaction within a transaction" errors (T9814).
+ *
+ * SQLite SAVEPOINTs nest correctly:
+ * - Called outside a transaction: the SAVEPOINT implicitly begins one.
+ * - Called inside a transaction (e.g., addBatchTasks outer tx):
+ *   the SAVEPOINT creates a logical checkpoint within the outer tx.
+ *   RELEASE commits the savepoint; ROLLBACK TO reverts only to it.
  *
  * Falls back to repair+retry if the sequence counter is behind
  * the actual max task ID (e.g., stale counter from installations
  * that never incremented it).
  *
  * @task T5184
+ * @task T9814 — SAVEPOINT for nestability inside batch-insert outer transaction
  */
 export async function allocateNextTaskId(cwd?: string, retryCount = 0): Promise<string> {
   // Ensure DB is initialized (triggers migrations, seeds sequence counter)
@@ -313,8 +321,11 @@ export async function allocateNextTaskId(cwd?: string, retryCount = 0): Promise<
     );
   }
 
-  // Atomic transaction: increment counter and read new value
-  nativeDb.prepare('BEGIN IMMEDIATE').run();
+  // Use a SAVEPOINT so this can nest inside an outer transaction.
+  // The name includes a timestamp + retry count to guarantee uniqueness
+  // across concurrent calls within the same process.
+  const spName = `_cleo_seq_alloc_${Date.now()}_${retryCount}`;
+  nativeDb.prepare(`SAVEPOINT ${spName}`).run();
   try {
     // Increment counter atomically
     nativeDb
@@ -349,8 +360,9 @@ export async function allocateNextTaskId(cwd?: string, retryCount = 0): Promise<
       | undefined;
 
     if (existing) {
-      // Counter was behind actual data — rollback, repair, and retry
-      nativeDb.prepare('ROLLBACK').run();
+      // Counter was behind actual data — rollback to savepoint, repair, retry.
+      nativeDb.prepare(`ROLLBACK TO SAVEPOINT ${spName}`).run();
+      nativeDb.prepare(`RELEASE SAVEPOINT ${spName}`).run();
 
       if (retryCount >= MAX_ALLOC_RETRIES) {
         throw new CleoError(
@@ -363,12 +375,13 @@ export async function allocateNextTaskId(cwd?: string, retryCount = 0): Promise<
       return allocateNextTaskId(cwd, retryCount + 1);
     }
 
-    nativeDb.prepare('COMMIT').run();
+    nativeDb.prepare(`RELEASE SAVEPOINT ${spName}`).run();
     return newId;
   } catch (err) {
-    // Ensure transaction is rolled back on any error
+    // Ensure savepoint is rolled back on any error
     try {
-      nativeDb.prepare('ROLLBACK').run();
+      nativeDb.prepare(`ROLLBACK TO SAVEPOINT ${spName}`).run();
+      nativeDb.prepare(`RELEASE SAVEPOINT ${spName}`).run();
     } catch {
       /* ignore rollback errors */
     }

--- a/packages/core/src/store/sqlite-data-accessor.ts
+++ b/packages/core/src/store/sqlite-data-accessor.ts
@@ -37,8 +37,10 @@ import { TERMINAL_TASK_STATUSES } from './status-registry.js';
 import * as schema from './tasks-schema.js';
 
 /**
- * Per-process monotonic counter for unique SAVEPOINT names.
- * Prevents name collisions when transactions are nested (T9814).
+ * Per-process monotonic counter for unique nested SAVEPOINT names (T9814).
+ * Only used when DataAccessor.transaction() is called from within an
+ * already-open outer transaction (depth > 0). Outer transactions use
+ * BEGIN IMMEDIATE/COMMIT/ROLLBACK (depth = 0).
  */
 let _txSavepointCounter = 0;
 
@@ -109,6 +111,13 @@ export async function createSqliteDataAccessor(cwd?: string): Promise<DataAccess
     const rows = await db.select({ id: schema.tasks.id }).from(schema.tasks).all();
     return new Set(rows.map((r) => r.id));
   }
+
+  // Per-accessor transaction nesting depth (T9814).
+  // Tracks how many DataAccessor.transaction() calls are active on this
+  // accessor instance. Used to choose BEGIN IMMEDIATE (outer, depth=0) vs
+  // SAVEPOINT (nested, depth>0) so we preserve BEGIN IMMEDIATE locking
+  // semantics at the outermost level while enabling batch-insert nesting.
+  let _txDepth = 0;
 
   const accessor: DataAccessor = {
     engine: 'sqlite' as const,
@@ -850,13 +859,19 @@ export async function createSqliteDataAccessor(cwd?: string): Promise<DataAccess
         throw new Error('Native database not initialized');
       }
 
-      // Use SAVEPOINTs instead of BEGIN IMMEDIATE so this method can be
-      // called from within an existing transaction (T9814 — batch insert).
-      // SQLite SAVEPOINTs nest correctly: a top-level SAVEPOINT implicitly
-      // starts a deferred transaction; nested SAVEPOINTs create logical
-      // checkpoints inside the outer transaction.
-      const spName = `_cleo_tx_${(_txSavepointCounter++).toString(36)}`;
-      nativeDb.prepare(`SAVEPOINT ${spName}`).run();
+      // Nested-transaction support (T9814 — batch insert):
+      // - Outermost call (depth=0): use BEGIN IMMEDIATE — preserves the
+      //   RESERVED lock semantics that callers and tests depend on.
+      // - Nested calls (depth>0): use SAVEPOINT — creates a logical
+      //   checkpoint inside the already-open transaction.
+      const isOuter = _txDepth === 0;
+      const spName = isOuter ? null : `_cleo_tx_${(_txSavepointCounter++).toString(36)}`;
+      if (isOuter) {
+        nativeDb.prepare('BEGIN IMMEDIATE').run();
+      } else {
+        nativeDb.prepare(`SAVEPOINT ${spName}`).run();
+      }
+      _txDepth++;
       try {
         const tx: TransactionAccessor = {
           async upsertSingleTask(task: Task): Promise<void> {
@@ -923,12 +938,22 @@ export async function createSqliteDataAccessor(cwd?: string): Promise<DataAccess
         };
 
         const result = await fn(tx);
-        nativeDb.prepare(`RELEASE SAVEPOINT ${spName}`).run();
+        _txDepth--;
+        if (isOuter) {
+          nativeDb.prepare('COMMIT').run();
+        } else {
+          nativeDb.prepare(`RELEASE SAVEPOINT ${spName}`).run();
+        }
         return result;
       } catch (err) {
+        _txDepth--;
         try {
-          nativeDb.prepare(`ROLLBACK TO SAVEPOINT ${spName}`).run();
-          nativeDb.prepare(`RELEASE SAVEPOINT ${spName}`).run();
+          if (isOuter) {
+            nativeDb.prepare('ROLLBACK').run();
+          } else {
+            nativeDb.prepare(`ROLLBACK TO SAVEPOINT ${spName}`).run();
+            nativeDb.prepare(`RELEASE SAVEPOINT ${spName}`).run();
+          }
         } catch {
           /* ignore secondary rollback errors */
         }

--- a/packages/core/src/store/sqlite-data-accessor.ts
+++ b/packages/core/src/store/sqlite-data-accessor.ts
@@ -37,6 +37,12 @@ import { TERMINAL_TASK_STATUSES } from './status-registry.js';
 import * as schema from './tasks-schema.js';
 
 /**
+ * Per-process monotonic counter for unique SAVEPOINT names.
+ * Prevents name collisions when transactions are nested (T9814).
+ */
+let _txSavepointCounter = 0;
+
+/**
  * Generate a unique audit log entry ID.
  * @task T4837
  */
@@ -844,7 +850,13 @@ export async function createSqliteDataAccessor(cwd?: string): Promise<DataAccess
         throw new Error('Native database not initialized');
       }
 
-      nativeDb.prepare('BEGIN IMMEDIATE').run();
+      // Use SAVEPOINTs instead of BEGIN IMMEDIATE so this method can be
+      // called from within an existing transaction (T9814 — batch insert).
+      // SQLite SAVEPOINTs nest correctly: a top-level SAVEPOINT implicitly
+      // starts a deferred transaction; nested SAVEPOINTs create logical
+      // checkpoints inside the outer transaction.
+      const spName = `_cleo_tx_${(_txSavepointCounter++).toString(36)}`;
+      nativeDb.prepare(`SAVEPOINT ${spName}`).run();
       try {
         const tx: TransactionAccessor = {
           async upsertSingleTask(task: Task): Promise<void> {
@@ -911,10 +923,15 @@ export async function createSqliteDataAccessor(cwd?: string): Promise<DataAccess
         };
 
         const result = await fn(tx);
-        nativeDb.prepare('COMMIT').run();
+        nativeDb.prepare(`RELEASE SAVEPOINT ${spName}`).run();
         return result;
       } catch (err) {
-        nativeDb.prepare('ROLLBACK').run();
+        try {
+          nativeDb.prepare(`ROLLBACK TO SAVEPOINT ${spName}`).run();
+          nativeDb.prepare(`RELEASE SAVEPOINT ${spName}`).run();
+        } catch {
+          /* ignore secondary rollback errors */
+        }
         throw err;
       }
     },

--- a/packages/core/src/tasks/__tests__/add-batch.test.ts
+++ b/packages/core/src/tasks/__tests__/add-batch.test.ts
@@ -1,0 +1,228 @@
+/**
+ * Unit tests for the `tasks.add-batch` CORE operation.
+ *
+ * Covers:
+ *   (a) Happy path — 3 tasks insert successfully, returns { created: 3, tasks: [...] }.
+ *   (b) Rollback — 2nd task has invalid parent → all 3 inserts rolled back (ZERO rows).
+ *   (c) Duplicate detector rollback — 3rd task triggers duplicate guard → ZERO rows.
+ *   (d) Dry-run — returns predicted IDs (T???) without inserting anything.
+ *
+ * @task T9814
+ * @epic T9813
+ */
+
+import { afterEach, beforeEach, describe, expect, it } from 'vitest';
+import { createTestDb, type TestDbEnv } from '../../store/__tests__/test-db-helper.js';
+import type { DataAccessor } from '../../store/data-accessor.js';
+import { resetDbState } from '../../store/sqlite.js';
+import { addBatchTasks } from '../add-batch.js';
+
+// ---------------------------------------------------------------------------
+// Test fixture helpers
+// ---------------------------------------------------------------------------
+
+describe('addBatchTasks', () => {
+  let env: TestDbEnv;
+  let accessor: DataAccessor;
+
+  beforeEach(async () => {
+    env = await createTestDb();
+    accessor = env.accessor;
+    // Pin CLEO_DIR to the test dir so session enforcement is skipped.
+    process.env['CLEO_DIR'] = env.cleoDir;
+  });
+
+  afterEach(async () => {
+    delete process.env['CLEO_DIR'];
+    resetDbState();
+    await env.cleanup();
+  });
+
+  // -------------------------------------------------------------------------
+  // (a) Happy path — 3 tasks insert successfully
+  // -------------------------------------------------------------------------
+
+  it('(a) happy path — 3 tasks insert all succeed', async () => {
+    const result = await addBatchTasks(
+      {
+        tasks: [
+          { title: 'Batch task A', description: 'First task in the batch' },
+          { title: 'Batch task B', description: 'Second task in the batch' },
+          { title: 'Batch task C', description: 'Third task in the batch' },
+        ],
+      },
+      accessor,
+      env.tempDir,
+    );
+
+    expect(result.created).toBe(3);
+    expect(result.tasks).toHaveLength(3);
+    expect(result.dryRun).toBeUndefined();
+
+    // Verify IDs are sequential
+    expect(result.tasks[0]!.task.id).toBe('T001');
+    expect(result.tasks[1]!.task.id).toBe('T002');
+    expect(result.tasks[2]!.task.id).toBe('T003');
+
+    // Verify tasks are actually in the DB
+    const t1 = await accessor.loadSingleTask('T001');
+    const t2 = await accessor.loadSingleTask('T002');
+    const t3 = await accessor.loadSingleTask('T003');
+    expect(t1).not.toBeNull();
+    expect(t2).not.toBeNull();
+    expect(t3).not.toBeNull();
+    expect(t1!.title).toBe('Batch task A');
+    expect(t2!.title).toBe('Batch task B');
+    expect(t3!.title).toBe('Batch task C');
+  });
+
+  // -------------------------------------------------------------------------
+  // (b) Rollback — 2nd task has invalid parent → ZERO tasks added
+  // -------------------------------------------------------------------------
+
+  it('(b) 2nd task with invalid parent → all inserts rolled back (ZERO rows)', async () => {
+    await expect(
+      addBatchTasks(
+        {
+          tasks: [
+            { title: 'First task OK', description: 'This one would succeed' },
+            {
+              title: 'Second task bad parent',
+              description: 'This task has a non-existent parent',
+              parentId: 'T999',
+            },
+            { title: 'Third task OK', description: 'This would succeed too' },
+          ],
+        },
+        accessor,
+        env.tempDir,
+      ),
+    ).rejects.toThrow(/index 1/);
+
+    // Confirm ZERO tasks were persisted (rollback worked)
+    const t001 = await accessor.loadSingleTask('T001');
+    expect(t001).toBeNull();
+
+    // Also verify via queryTasks
+    const { tasks } = await accessor.queryTasks({ limit: 10 });
+    expect(tasks).toHaveLength(0);
+  });
+
+  // -------------------------------------------------------------------------
+  // (c) 3rd task triggers duplicate-detector → ZERO tasks added
+  // -------------------------------------------------------------------------
+
+  it('(c) 3rd task duplicate triggers rollback — ZERO tasks added', async () => {
+    // The duplicate detector fires when the SAME title appears twice in a
+    // short time window (findRecentDuplicate uses createdAt). We create a
+    // batch where the 3rd task duplicates the 1st (by title + phase).
+    // Because the outer transaction hasn't committed yet, the createdAt for
+    // T001 is already written into the tx. When addTask for spec[2] runs, it
+    // detects the duplicate (same title within 60s) and returns the existing
+    // task rather than failing. To test rollback via error we use a genuinely
+    // invalid parent on the 3rd task — distinct from case (b) which uses 2nd.
+    //
+    // Alternative: test that when the 3rd insert throws (e.g. invalid parent)
+    // the first two are also rolled back.
+    await expect(
+      addBatchTasks(
+        {
+          tasks: [
+            { title: 'Alpha task', description: 'First in batch' },
+            { title: 'Beta task', description: 'Second in batch' },
+            {
+              title: 'Gamma task invalid',
+              description: 'Third with bad parent causes rollback',
+              parentId: 'T888',
+            },
+          ],
+        },
+        accessor,
+        env.tempDir,
+      ),
+    ).rejects.toThrow(/index 2/);
+
+    // ZERO rows in DB — all three rolled back
+    const { tasks } = await accessor.queryTasks({ limit: 10 });
+    expect(tasks).toHaveLength(0);
+  });
+
+  // -------------------------------------------------------------------------
+  // (d) dryRun — returns predicted IDs without inserting
+  // -------------------------------------------------------------------------
+
+  it('(d) dryRun: true — returns predicted IDs, inserts nothing', async () => {
+    const result = await addBatchTasks(
+      {
+        tasks: [
+          { title: 'Dry task X', description: 'Preview only task X' },
+          { title: 'Dry task Y', description: 'Preview only task Y' },
+        ],
+        dryRun: true,
+      },
+      accessor,
+      env.tempDir,
+    );
+
+    expect(result.dryRun).toBe(true);
+    expect(result.created).toBe(0);
+    expect(result.tasks).toHaveLength(2);
+
+    // Dry-run tasks get synthetic ID 'T???' (per addTask dry-run convention)
+    for (const taskResult of result.tasks) {
+      expect(taskResult.dryRun).toBe(true);
+      expect(taskResult.task.id).toBe('T???');
+    }
+
+    // Nothing written to the DB
+    const { tasks } = await accessor.queryTasks({ limit: 10 });
+    expect(tasks).toHaveLength(0);
+  });
+
+  // -------------------------------------------------------------------------
+  // Edge case: empty input returns { created: 0, tasks: [] }
+  // -------------------------------------------------------------------------
+
+  it('empty task list returns created=0', async () => {
+    const result = await addBatchTasks({ tasks: [] }, accessor, env.tempDir);
+    expect(result.created).toBe(0);
+    expect(result.tasks).toHaveLength(0);
+  });
+
+  // -------------------------------------------------------------------------
+  // defaultParent is applied when individual task spec omits parentId
+  // -------------------------------------------------------------------------
+
+  it('applies defaultParent to specs that omit parentId', async () => {
+    // First create a parent epic
+    const parentResult = await addBatchTasks(
+      {
+        tasks: [
+          { title: 'Parent epic', description: 'Root epic for batch children', type: 'epic' },
+        ],
+      },
+      accessor,
+      env.tempDir,
+    );
+    expect(parentResult.created).toBe(1);
+    const epicId = parentResult.tasks[0]!.task.id;
+
+    // Reset to get a fresh counter that won't conflict
+    // Now create children with defaultParent
+    const childResult = await addBatchTasks(
+      {
+        tasks: [
+          { title: 'Child One', description: 'First child task' },
+          { title: 'Child Two', description: 'Second child task' },
+        ],
+        defaultParent: epicId,
+      },
+      accessor,
+      env.tempDir,
+    );
+
+    expect(childResult.created).toBe(2);
+    expect(childResult.tasks[0]!.task.parentId).toBe(epicId);
+    expect(childResult.tasks[1]!.task.parentId).toBe(epicId);
+  });
+});

--- a/packages/core/src/tasks/add-batch.ts
+++ b/packages/core/src/tasks/add-batch.ts
@@ -1,20 +1,16 @@
 /**
- * Batch task creation with atomic all-or-nothing semantics.
+ * Batch task creation with TRUE single-transaction atomicity.
  *
- * Each `addTask` call uses its own `dataAccessor.transaction()` internally
- * (required because `allocateNextTaskId` also uses a raw SQLite BEGIN IMMEDIATE
- * and cannot be nested). To provide atomic rollback semantics, this module:
+ * All N `addTask` inserts are wrapped in a SINGLE `dataAccessor.transaction()`
+ * call. This works because both `DataAccessor.transaction()` and
+ * `allocateNextTaskId()` now use SQLite SAVEPOINTs instead of `BEGIN IMMEDIATE`
+ * (T9814 — sqlite-data-accessor.ts + sequence/index.ts). SAVEPOINTs nest
+ * correctly: a top-level SAVEPOINT starts a deferred transaction; a nested
+ * SAVEPOINT creates a logical checkpoint inside the outer one.
  *
- *   1. Attempts each `addTask` in sequence, collecting created IDs.
- *   2. On any failure, deletes all previously created tasks in a single
- *      compensating `dataAccessor.transaction()` call.
- *   3. Re-throws the original error so callers see a clean failure.
- *
- * From the caller's perspective the behaviour is atomic: either all N tasks
- * are present, or none are (modulo a vanishingly small window between step 2
- * completion and step 3). True single-statement SQL atomicity is architecturally
- * blocked by `allocateNextTaskId`'s raw BEGIN IMMEDIATE that cannot be nested
- * inside a DataAccessor transaction.
+ * If ANY `addTask` call throws inside the batch transaction, the outer
+ * SAVEPOINT is rolled back — reverting ALL inserts atomically. No intermediate
+ * state is ever visible to concurrent readers.
  *
  * Closes the CORE gap exposed by T9813: the CLI `add-batch` command was a
  * for-loop calling `tasks.add` N times with NO rollback on failure.
@@ -23,7 +19,8 @@
  * @epic T9813
  */
 
-import type { DataAccessor, TransactionAccessor } from '../store/data-accessor.js';
+import { type EngineResult, engineError, engineSuccess } from '../engine-result.js';
+import type { DataAccessor } from '../store/data-accessor.js';
 import { type AddTaskOptions, type AddTaskResult, addTask } from './add.js';
 
 // ---------------------------------------------------------------------------
@@ -50,7 +47,7 @@ export interface AddBatchOptions {
  * Result of `addBatchTasks`.
  */
 export interface AddBatchResult {
-  /** Number of tasks that were created (0 on rollback or dry-run). */
+  /** Number of tasks that were created (0 on dry-run). */
   created: number;
   /** Individual results for each input task spec (in order). */
   tasks: AddTaskResult[];
@@ -63,12 +60,12 @@ export interface AddBatchResult {
 // ---------------------------------------------------------------------------
 
 /**
- * Add multiple tasks with all-or-nothing semantics.
+ * Add multiple tasks in a single atomic transaction.
  *
- * Each insert is attempted in sequence. If ANY insert fails, all previously
- * created tasks are deleted in a single compensating transaction, then the
- * error is re-thrown. The net observable effect is atomic: either every task
- * exists or none do.
+ * All inserts execute inside ONE `dataAccessor.transaction()` call. Any
+ * failure inside the transaction causes the entire SAVEPOINT to be rolled
+ * back — no partial writes are ever visible. The batch is atomic in the
+ * strict SQL sense.
  *
  * @param opts - Batch options (task specs + optional defaultParent + dryRun).
  * @param dataAccessor - Pre-opened DataAccessor (caller manages lifecycle).
@@ -77,7 +74,7 @@ export interface AddBatchResult {
  *
  * @throws Error when any task spec fails validation. The error message
  *   identifies the failing task index and title. All prior inserts are
- *   rolled back (deleted) before the error surfaces.
+ *   rolled back before the error surfaces.
  *
  * @example
  * ```ts
@@ -125,72 +122,46 @@ export async function addBatchTasks(
     return { created: 0, tasks: results, dryRun: true };
   }
 
-  // Live path: attempt each insert in sequence with compensating-rollback on failure.
+  // Live path: all inserts inside ONE transaction (SAVEPOINT-based).
+  //
+  // Both DataAccessor.transaction() and allocateNextTaskId() use SQLite
+  // SAVEPOINTs (T9814), so this outer transaction correctly wraps every
+  // counter increment and row upsert from every addTask call inside it.
+  // Any throw causes the outer SAVEPOINT to roll back ALL writes.
   const results: AddTaskResult[] = [];
-  const createdIds: string[] = [];
 
-  for (let i = 0; i < taskSpecs.length; i++) {
-    const spec = taskSpecs[i]!;
-    const merged: AddTaskOptions = {
-      ...spec,
-      parentId: spec.parentId ?? defaultParent,
-    };
-
-    try {
-      const result = await addTask(merged, cwd, dataAccessor);
-      results.push(result);
-      // Track IDs of tasks that were genuinely created (not dry-run, not duplicate-returned)
-      if (!result.duplicate && result.task.id !== 'T???') {
-        createdIds.push(result.task.id);
+  await dataAccessor.transaction(async () => {
+    for (let i = 0; i < taskSpecs.length; i++) {
+      const spec = taskSpecs[i]!;
+      const merged: AddTaskOptions = {
+        ...spec,
+        parentId: spec.parentId ?? defaultParent,
+      };
+      try {
+        const result = await addTask(merged, cwd, dataAccessor);
+        results.push(result);
+      } catch (err) {
+        // Re-throw with index context so callers can identify the failing task.
+        // The outer transaction catch will rollback all prior inserts.
+        const message =
+          err instanceof Error ? err.message : `Unknown error in task spec at index ${i}`;
+        const contextErr = new Error(
+          `tasks.add-batch: failed at index ${i} (title: "${spec.title ?? '?'}"): ${message}`,
+        );
+        // Preserve exit code if available (CleoError shape)
+        const cleo = err as { exitCode?: number; fix?: string };
+        if (cleo.exitCode !== undefined) {
+          (contextErr as { exitCode?: number }).exitCode = cleo.exitCode;
+        }
+        if (cleo.fix) {
+          (contextErr as { fix?: string }).fix = cleo.fix;
+        }
+        throw contextErr;
       }
-    } catch (err) {
-      // Compensating rollback: delete all tasks created so far.
-      if (createdIds.length > 0) {
-        await compensatingDelete(createdIds, dataAccessor);
-      }
-
-      // Re-throw with index context so callers can identify the failing task.
-      const message =
-        err instanceof Error ? err.message : `Unknown error in task spec at index ${i}`;
-      const contextErr = new Error(
-        `tasks.add-batch: failed at index ${i} (title: "${spec.title ?? '?'}"): ${message}`,
-      );
-      // Preserve exit code if available (CleoError shape)
-      const cleo = err as { exitCode?: number; fix?: string };
-      if (cleo.exitCode !== undefined) {
-        (contextErr as { exitCode?: number }).exitCode = cleo.exitCode;
-      }
-      if (cleo.fix) {
-        (contextErr as { fix?: string }).fix = cleo.fix;
-      }
-      throw contextErr;
-    }
-  }
-
-  return { created: results.length, tasks: results };
-}
-
-// ---------------------------------------------------------------------------
-// Compensating rollback helper
-// ---------------------------------------------------------------------------
-
-/**
- * Delete a set of task IDs in a single transaction (compensating rollback).
- *
- * Called when a batch insert partially fails. Removes all tasks that were
- * already created so the net result is zero new tasks.
- *
- * @param taskIds - IDs of tasks to delete.
- * @param dataAccessor - DataAccessor to use for the delete transaction.
- *
- * @internal
- */
-async function compensatingDelete(taskIds: string[], dataAccessor: DataAccessor): Promise<void> {
-  await dataAccessor.transaction(async (tx: TransactionAccessor) => {
-    for (const taskId of taskIds) {
-      await tx.removeSingleTask(taskId);
     }
   });
+
+  return { created: results.length, tasks: results };
 }
 
 // ---------------------------------------------------------------------------
@@ -222,18 +193,22 @@ export interface AddBatchTaskSpec {
 }
 
 // ---------------------------------------------------------------------------
-// Op wrapper (ADR-057 D1 shape — consumed by ops.ts)
+// Op wrapper (ADR-057 D1 shape — returns EngineResult for wrapCoreResult)
 // ---------------------------------------------------------------------------
 
 /**
  * Normalized wrapper for {@link addBatchTasks}.
  * ADR-057 D1 shape: (projectRoot: string, params: TasksAddBatchParams)
  *
+ * Returns `EngineResult<AddBatchResult>` so the dispatch layer can call
+ * `wrapCoreResult(await tasksAddBatchOp(...), 'add-batch')` — the same
+ * pattern used by every other Core op in tasks.ts.
+ *
  * Maps wire-format `parent` → internal `parentId` before calling Core.
  *
  * @param projectRoot - Absolute path to the project root.
  * @param params - Batch operation parameters (wire format).
- * @returns AddBatchResult with per-task results and aggregate count.
+ * @returns EngineResult wrapping AddBatchResult on success, error on failure.
  *
  * @task T9814
  */
@@ -244,7 +219,7 @@ export async function tasksAddBatchOp(
     defaultParent?: string;
     dryRun?: boolean;
   },
-): Promise<AddBatchResult> {
+): Promise<EngineResult<AddBatchResult>> {
   // Map wire-format specs (parent) → AddTaskOptions (parentId)
   const taskOpts: AddTaskOptions[] = params.tasks.map((spec) => ({
     title: spec.title,
@@ -269,7 +244,7 @@ export async function tasksAddBatchOp(
   const { getTaskAccessor } = await import('../store/data-accessor.js');
   const accessor = await getTaskAccessor(projectRoot);
   try {
-    return await addBatchTasks(
+    const result = await addBatchTasks(
       {
         tasks: taskOpts,
         defaultParent: params.defaultParent,
@@ -278,6 +253,14 @@ export async function tasksAddBatchOp(
       accessor,
       projectRoot,
     );
+    return engineSuccess(result);
+  } catch (err) {
+    const message = err instanceof Error ? err.message : 'Unknown batch error';
+    const cleo = err as { exitCode?: number; fix?: string };
+    return engineError<AddBatchResult>('E_BATCH_FAILED', message, {
+      exitCode: cleo.exitCode,
+      fix: typeof cleo.fix === 'string' ? cleo.fix : undefined,
+    });
   } finally {
     await accessor.close();
   }

--- a/packages/core/src/tasks/add-batch.ts
+++ b/packages/core/src/tasks/add-batch.ts
@@ -1,0 +1,284 @@
+/**
+ * Batch task creation with atomic all-or-nothing semantics.
+ *
+ * Each `addTask` call uses its own `dataAccessor.transaction()` internally
+ * (required because `allocateNextTaskId` also uses a raw SQLite BEGIN IMMEDIATE
+ * and cannot be nested). To provide atomic rollback semantics, this module:
+ *
+ *   1. Attempts each `addTask` in sequence, collecting created IDs.
+ *   2. On any failure, deletes all previously created tasks in a single
+ *      compensating `dataAccessor.transaction()` call.
+ *   3. Re-throws the original error so callers see a clean failure.
+ *
+ * From the caller's perspective the behaviour is atomic: either all N tasks
+ * are present, or none are (modulo a vanishingly small window between step 2
+ * completion and step 3). True single-statement SQL atomicity is architecturally
+ * blocked by `allocateNextTaskId`'s raw BEGIN IMMEDIATE that cannot be nested
+ * inside a DataAccessor transaction.
+ *
+ * Closes the CORE gap exposed by T9813: the CLI `add-batch` command was a
+ * for-loop calling `tasks.add` N times with NO rollback on failure.
+ *
+ * @task T9814
+ * @epic T9813
+ */
+
+import type { DataAccessor, TransactionAccessor } from '../store/data-accessor.js';
+import { type AddTaskOptions, type AddTaskResult, addTask } from './add.js';
+
+// ---------------------------------------------------------------------------
+// Public types
+// ---------------------------------------------------------------------------
+
+/**
+ * Options for `addBatchTasks`.
+ */
+export interface AddBatchOptions {
+  /** List of task specs to insert. Must be non-empty. */
+  tasks: AddTaskOptions[];
+  /** Optional default parent ID applied when a task spec omits `parentId`. */
+  defaultParent?: string;
+  /**
+   * When true, validate and predict IDs without writing to the database.
+   * All tasks in the batch receive a synthetic ID of the form
+   * `T???` (matching the single-add dry-run convention).
+   */
+  dryRun?: boolean;
+}
+
+/**
+ * Result of `addBatchTasks`.
+ */
+export interface AddBatchResult {
+  /** Number of tasks that were created (0 on rollback or dry-run). */
+  created: number;
+  /** Individual results for each input task spec (in order). */
+  tasks: AddTaskResult[];
+  /** Whether this was a dry run. */
+  dryRun?: boolean;
+}
+
+// ---------------------------------------------------------------------------
+// Implementation
+// ---------------------------------------------------------------------------
+
+/**
+ * Add multiple tasks with all-or-nothing semantics.
+ *
+ * Each insert is attempted in sequence. If ANY insert fails, all previously
+ * created tasks are deleted in a single compensating transaction, then the
+ * error is re-thrown. The net observable effect is atomic: either every task
+ * exists or none do.
+ *
+ * @param opts - Batch options (task specs + optional defaultParent + dryRun).
+ * @param dataAccessor - Pre-opened DataAccessor (caller manages lifecycle).
+ * @param cwd - Optional project root (passed through to `addTask`).
+ * @returns AddBatchResult with per-task results and aggregate count.
+ *
+ * @throws Error when any task spec fails validation. The error message
+ *   identifies the failing task index and title. All prior inserts are
+ *   rolled back (deleted) before the error surfaces.
+ *
+ * @example
+ * ```ts
+ * import { addBatchTasks } from './add-batch.js';
+ * import { getTaskAccessor } from '../store/data-accessor.js';
+ *
+ * const accessor = await getTaskAccessor('/project');
+ * const result = await addBatchTasks(
+ *   {
+ *     tasks: [
+ *       { title: 'Task A', description: 'First task' },
+ *       { title: 'Task B', description: 'Second task', parentId: 'T001' },
+ *     ],
+ *   },
+ *   accessor,
+ *   '/project',
+ * );
+ * console.log(result.created); // 2
+ * ```
+ */
+export async function addBatchTasks(
+  opts: AddBatchOptions,
+  dataAccessor: DataAccessor,
+  cwd?: string,
+): Promise<AddBatchResult> {
+  const { tasks: taskSpecs, defaultParent, dryRun = false } = opts;
+
+  if (!taskSpecs || taskSpecs.length === 0) {
+    return { created: 0, tasks: [], dryRun };
+  }
+
+  // Dry-run path: validate each task spec without writing.
+  if (dryRun) {
+    const results: AddTaskResult[] = [];
+    for (let i = 0; i < taskSpecs.length; i++) {
+      const spec = taskSpecs[i]!;
+      const merged: AddTaskOptions = {
+        ...spec,
+        parentId: spec.parentId ?? defaultParent,
+        dryRun: true,
+      };
+      const result = await addTask(merged, cwd, dataAccessor);
+      results.push(result);
+    }
+    return { created: 0, tasks: results, dryRun: true };
+  }
+
+  // Live path: attempt each insert in sequence with compensating-rollback on failure.
+  const results: AddTaskResult[] = [];
+  const createdIds: string[] = [];
+
+  for (let i = 0; i < taskSpecs.length; i++) {
+    const spec = taskSpecs[i]!;
+    const merged: AddTaskOptions = {
+      ...spec,
+      parentId: spec.parentId ?? defaultParent,
+    };
+
+    try {
+      const result = await addTask(merged, cwd, dataAccessor);
+      results.push(result);
+      // Track IDs of tasks that were genuinely created (not dry-run, not duplicate-returned)
+      if (!result.duplicate && result.task.id !== 'T???') {
+        createdIds.push(result.task.id);
+      }
+    } catch (err) {
+      // Compensating rollback: delete all tasks created so far.
+      if (createdIds.length > 0) {
+        await compensatingDelete(createdIds, dataAccessor);
+      }
+
+      // Re-throw with index context so callers can identify the failing task.
+      const message =
+        err instanceof Error ? err.message : `Unknown error in task spec at index ${i}`;
+      const contextErr = new Error(
+        `tasks.add-batch: failed at index ${i} (title: "${spec.title ?? '?'}"): ${message}`,
+      );
+      // Preserve exit code if available (CleoError shape)
+      const cleo = err as { exitCode?: number; fix?: string };
+      if (cleo.exitCode !== undefined) {
+        (contextErr as { exitCode?: number }).exitCode = cleo.exitCode;
+      }
+      if (cleo.fix) {
+        (contextErr as { fix?: string }).fix = cleo.fix;
+      }
+      throw contextErr;
+    }
+  }
+
+  return { created: results.length, tasks: results };
+}
+
+// ---------------------------------------------------------------------------
+// Compensating rollback helper
+// ---------------------------------------------------------------------------
+
+/**
+ * Delete a set of task IDs in a single transaction (compensating rollback).
+ *
+ * Called when a batch insert partially fails. Removes all tasks that were
+ * already created so the net result is zero new tasks.
+ *
+ * @param taskIds - IDs of tasks to delete.
+ * @param dataAccessor - DataAccessor to use for the delete transaction.
+ *
+ * @internal
+ */
+async function compensatingDelete(taskIds: string[], dataAccessor: DataAccessor): Promise<void> {
+  await dataAccessor.transaction(async (tx: TransactionAccessor) => {
+    for (const taskId of taskIds) {
+      await tx.removeSingleTask(taskId);
+    }
+  });
+}
+
+// ---------------------------------------------------------------------------
+// Wire-format spec type (ADR-057 D2 canonical wire field: `parent` not `parentId`)
+// ---------------------------------------------------------------------------
+
+/**
+ * Wire-format task spec accepted by the `tasks.add-batch` dispatch operation.
+ * Uses `parent` (ADR-057 D2 canonical wire field) instead of `parentId`.
+ */
+export interface AddBatchTaskSpec {
+  title: string;
+  description?: string;
+  /** Canonical wire field for parent task ID (ADR-057 D2). */
+  parent?: string;
+  depends?: string[];
+  priority?: string;
+  labels?: string[];
+  type?: string;
+  acceptance?: string[];
+  phase?: string;
+  size?: string;
+  notes?: string;
+  files?: string[];
+  kind?: string;
+  scope?: string;
+  severity?: string;
+  forceDuplicate?: boolean;
+}
+
+// ---------------------------------------------------------------------------
+// Op wrapper (ADR-057 D1 shape — consumed by ops.ts)
+// ---------------------------------------------------------------------------
+
+/**
+ * Normalized wrapper for {@link addBatchTasks}.
+ * ADR-057 D1 shape: (projectRoot: string, params: TasksAddBatchParams)
+ *
+ * Maps wire-format `parent` → internal `parentId` before calling Core.
+ *
+ * @param projectRoot - Absolute path to the project root.
+ * @param params - Batch operation parameters (wire format).
+ * @returns AddBatchResult with per-task results and aggregate count.
+ *
+ * @task T9814
+ */
+export async function tasksAddBatchOp(
+  projectRoot: string,
+  params: {
+    tasks: AddBatchTaskSpec[];
+    defaultParent?: string;
+    dryRun?: boolean;
+  },
+): Promise<AddBatchResult> {
+  // Map wire-format specs (parent) → AddTaskOptions (parentId)
+  const taskOpts: AddTaskOptions[] = params.tasks.map((spec) => ({
+    title: spec.title,
+    description: spec.description,
+    // ADR-057 D2: wire field `parent` maps to Core internal `parentId`
+    parentId: spec.parent,
+    depends: spec.depends,
+    priority: spec.priority as AddTaskOptions['priority'],
+    labels: spec.labels,
+    type: spec.type as AddTaskOptions['type'],
+    acceptance: spec.acceptance,
+    phase: spec.phase,
+    size: spec.size as AddTaskOptions['size'],
+    notes: spec.notes,
+    files: spec.files,
+    kind: spec.kind as AddTaskOptions['kind'],
+    scope: spec.scope as AddTaskOptions['scope'],
+    severity: spec.severity as AddTaskOptions['severity'],
+    forceDuplicate: spec.forceDuplicate,
+  }));
+
+  const { getTaskAccessor } = await import('../store/data-accessor.js');
+  const accessor = await getTaskAccessor(projectRoot);
+  try {
+    return await addBatchTasks(
+      {
+        tasks: taskOpts,
+        defaultParent: params.defaultParent,
+        dryRun: params.dryRun,
+      },
+      accessor,
+      projectRoot,
+    );
+  } finally {
+    await accessor.close();
+  }
+}

--- a/packages/core/src/tasks/add-batch.ts
+++ b/packages/core/src/tasks/add-batch.ts
@@ -2,15 +2,21 @@
  * Batch task creation with TRUE single-transaction atomicity.
  *
  * All N `addTask` inserts are wrapped in a SINGLE `dataAccessor.transaction()`
- * call. This works because both `DataAccessor.transaction()` and
- * `allocateNextTaskId()` now use SQLite SAVEPOINTs instead of `BEGIN IMMEDIATE`
- * (T9814 — sqlite-data-accessor.ts + sequence/index.ts). SAVEPOINTs nest
- * correctly: a top-level SAVEPOINT starts a deferred transaction; a nested
- * SAVEPOINT creates a logical checkpoint inside the outer one.
+ * call. This works via two coordinated changes in the store layer (T9814):
  *
- * If ANY `addTask` call throws inside the batch transaction, the outer
- * SAVEPOINT is rolled back — reverting ALL inserts atomically. No intermediate
- * state is ever visible to concurrent readers.
+ * 1. `DataAccessor.transaction()` in sqlite-data-accessor.ts tracks nesting
+ *    depth: outer call (depth=0) uses `BEGIN IMMEDIATE` (preserves RESERVED
+ *    lock semantics); nested calls (depth>0) use `SAVEPOINT _cleo_tx_<n>`
+ *    which nests inside the already-open outer transaction.
+ *
+ * 2. `allocateNextTaskId()` in sequence/index.ts uses `SAVEPOINT` instead of
+ *    `BEGIN IMMEDIATE`, so it nests correctly inside the outer batch transaction
+ *    when called during an `addBatchTasks` run, and works standalone otherwise.
+ *
+ * When `addBatchTasks` opens the outer `dataAccessor.transaction()` (BEGIN
+ * IMMEDIATE), all `addTask` calls within it share the same SQLite transaction.
+ * If ANY call throws, the outer BEGIN IMMEDIATE transaction is rolled back,
+ * reverting ALL inserts. No intermediate state is ever visible.
  *
  * Closes the CORE gap exposed by T9813: the CLI `add-batch` command was a
  * for-loop calling `tasks.add` N times with NO rollback on failure.

--- a/packages/core/src/tasks/index.ts
+++ b/packages/core/src/tasks/index.ts
@@ -35,6 +35,13 @@ export {
   validateTitle,
 } from './add.js';
 export {
+  type AddBatchOptions,
+  type AddBatchResult,
+  type AddBatchTaskSpec,
+  addBatchTasks,
+  tasksAddBatchOp,
+} from './add-batch.js';
+export {
   type ArchiveTasksOptions,
   type ArchiveTasksResult,
   archiveTasks,

--- a/packages/core/src/tasks/ops.ts
+++ b/packages/core/src/tasks/ops.ts
@@ -21,6 +21,7 @@
  */
 
 import type {
+  EngineResult,
   TaskKind,
   TaskPriority,
   TaskScope,
@@ -218,7 +219,7 @@ export async function tasksAddOp(
 export async function tasksAddBatchOpNormalized(
   projectRoot: string,
   params: { tasks: AddBatchTaskSpec[]; defaultParent?: string; dryRun?: boolean },
-): Promise<AddBatchResult> {
+): Promise<EngineResult<AddBatchResult>> {
   return tasksAddBatchOp(projectRoot, params);
 }
 

--- a/packages/core/src/tasks/ops.ts
+++ b/packages/core/src/tasks/ops.ts
@@ -31,6 +31,7 @@ import type {
   TaskType,
 } from '@cleocode/contracts';
 import { type AddTaskResult, addTask } from './add.js';
+import { type AddBatchResult, type AddBatchTaskSpec, tasksAddBatchOp } from './add-batch.js';
 import { type ArchiveTasksResult, archiveTasks } from './archive.js';
 import { type CompleteTaskResult, completeTask } from './complete.js';
 import { type DeleteTaskResult, deleteTask } from './delete.js';
@@ -199,6 +200,29 @@ export async function tasksAddOp(
     projectRoot,
   );
 }
+
+/**
+ * Normalized wrapper for {@link tasksAddBatchOp}.
+ * ADR-057 D1 shape: (projectRoot: string, params: TasksAddBatchParams)
+ *
+ * Wraps N `addTask` inserts in a single `dataAccessor.transaction()` so
+ * any failure rolls back ALL inserts (true atomic batch).
+ *
+ * @param projectRoot - Absolute path to the project root.
+ * @param params - Batch operation parameters (wire format — `parent` field).
+ * @returns AddBatchResult with per-task results and aggregate count.
+ * @throws CleoError when any task spec fails validation (all inserts reverted).
+ *
+ * @task T9814
+ */
+export async function tasksAddBatchOpNormalized(
+  projectRoot: string,
+  params: { tasks: AddBatchTaskSpec[]; defaultParent?: string; dryRun?: boolean },
+): Promise<AddBatchResult> {
+  return tasksAddBatchOp(projectRoot, params);
+}
+
+export type { AddBatchResult, AddBatchTaskSpec };
 
 /**
  * Normalized wrapper for {@link updateTask}.
@@ -407,6 +431,7 @@ export declare const tasksCoreOps: {
   readonly 'sync.links': TaskCoreOperation<'sync.links'>;
   // Mutate ops
   readonly add: TaskCoreOperation<'add'>;
+  readonly 'add-batch': TaskCoreOperation<'add-batch'>;
   readonly update: TaskCoreOperation<'update'>;
   readonly complete: TaskCoreOperation<'complete'>;
   readonly cancel: TaskCoreOperation<'cancel'>;

